File: Week_06/G20190343020335/implement-trie-prefix-tree.py
class TrieNode(dict):
    def __init__(self, *args, **kwargs):
        self.end = False
        super(TrieNode, self).__init__(*args, **kwargs)


class Trie(object):

    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        cur = self.root
        for char in word:
            node = cur.get(char)
            if node is None:
                node = TrieNode()
                cur[char] = node
            cur = node
        cur.end = True

    def search(self, word):
        cur = self.root
        for char in word:
            cur = cur.get(char)
            if cur is None:
                return False
        return cur.end

    def startsWith(self, prefix):
        cur = self.root
        for char in prefix:
            cur = cur.get(char)
            if cur is None:
                return False
        return True

File: Week_06/G20190343020023/implement-trie-prefix-tree.go
type Trie struct {
	isWordEnd bool
	next      [26]*Trie
}

/** Initialize your data structure here. */
func Constructor() Trie {
	return Trie{}
}

/** Inserts a word into the trie. */
func (this *Trie) Insert(word string) {
	for _, char := range word {
		if this.next[char-'a'] == nil {
			this.next[char-'a'] = new(Trie)
		}
		this = this.next[char-'a']
	}
	this.isWordEnd = true
}

/** Returns if the word is in the trie. */
func (this *Trie) Search(word string) bool {
	for _, char := range word {
		if this.next[char-'a'] == nil {
			return false
		}
		this = this.next[char-'a']
	}
	return this.isWordEnd
}

/** Returns if there is any word in the trie that starts with the given prefix. */
func (this *Trie) StartsWith(prefix string) bool {
	for _, char := range prefix {
		if this.next[char-'a'] == nil {
			return false
		}
		this = this.next[char-'a']
	}
	return true
}

/**
 * Your Trie object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Insert(word);
 * param_2 := obj.Search(word);
 * param_3 := obj.StartsWith(prefix);
 */